app/src/main/kotlin/com/simplemobiletools/keyboard/extensions/Context.kt

```diff
@@ -36,3 +45,107 @@ fun Context.getStrokeColor(): Int {
         }
     }
 }
+
+fun Context.getKeyboardDialogBuilder() = if (baseConfig.isUsingSystemTheme) {
+    MaterialAlertDialogBuilder(this, R.style.MyKeyboard_Alert)
+} else {
+    AlertDialog.Builder(this, R.style.MyKeyboard_Alert)
+}
+
+
+fun Context.setupKeyboardDialogStuff(
+    windowToken: IBinder,
+    view: View,
+    dialog: AlertDialog.Builder,
+    titleId: Int = 0,
+    titleText: String = "",
+    cancelOnTouchOutside: Boolean = true,
+    callback: ((alertDialog: AlertDialog) -> Unit)? = null
+) {
+
+
+    val textColor = getProperTextColor()
+    val backgroundColor = getProperBackgroundColor()
+    val primaryColor = getProperPrimaryColor()
+    if (view is ViewGroup) {
+        updateTextColors(view)
+    } else if (view is MyTextView) {
+        view.setColors(textColor, primaryColor, backgroundColor)
+    }
+
+    if (dialog is MaterialAlertDialogBuilder) {
+        dialog.create().apply {
+            if (titleId != 0) {
+                setTitle(titleId)
+            } else if (titleText.isNotEmpty()) {
+                setTitle(titleText)
+            }
+
+            val lp = window?.attributes
+            lp?.token = windowToken
```

Comment:
This was the part I missed initially. I thought it was impossible to show an actual dialog from a service. But in the case of the IME service, it has a window for the keyboard and we can reuse the token to show a dialog 😓